lua/apisix/plugins/batch-processor.lua

```diff
@@ -0,0 +1,180 @@
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core = require("apisix.core")
+local setmetatable = setmetatable
+local timer_at = ngx.timer.at
+local remove = table.remove
+local fmt = string.format
+local ngx_log = ngx.log
+local DEBUG = ngx.DEBUG
+local table = table
+local now = ngx.now
+local type = type
+local Batch_Processor = {}
+local Batch_Processor_mt = {
+    __index = Batch_Processor
+}
+local execute_func
+local create_buffer_timer
+
+
+local schema = {
+    type = "object",
+    properties = {
+        name = {type = "string", default = "log buffer"},
+        max_retry_count = {type = "integer", minimum = 0, default= 0},
+        retry_delay = {type = "integer", minimum = 0, default= 1},
+        buffer_duration = {type = "integer", minimum = 1, default= 60}, -- maximum age in seconds of the oldest log item in a batch before the batch must be transmitted
+        inactive_timeout = {type = "integer", minimum = 1, default= 5}, -- maximum age in seconds when the buffer will be flushed if inactive
+        batch_max_size = {type = "integer", minimum = 1, default= 1000}, -- maximum number of entries in a batch before the batch must be transmitted
+    }
+}
+
+
+local function schedule_func_exec(batch_processor, delay, batch)
+    local hdl, err = timer_at(delay, execute_func, batch_processor, batch)
+    if not hdl then
+        core.log.error("failed to create process timer: ", err)
+        return
+    end
+end
+
+
+execute_func = function(premature, batch_processor, batch)
+    if premature then
+        return
+    end
+
+    local ok, err = batch_processor.func(batch.entries)
+    if ok then
+        ngx_log(DEBUG, fmt("Batch Processor[%s] successfully processed the entries", batch_processor.name))
+
+    else
+        batch.retry_count = batch.retry_count + 1
+        if batch.retry_count < batch_processor.max_retry_count then
+            core.log.warn(fmt("Batch Processor[%s] failed to process entries: ", batch_processor.name), err)
+            schedule_func_exec(batch_processor, batch_processor.retry_delay, batch)
+        else
+            core.log.error(fmt("Batch Processor[%s] exceeded the max_retry_count[%d], dropping the entries", batch_processor.name, batch.retry_count))
+        end
+    end
+end
+
+
+local function flush_buffer(premature, batch_processor)
+    if premature then
+        return
+    end
+
+    if now() - batch_processor.last_entry_t >= batch_processor.inactive_timeout or
+            now() - batch_processor.first_entry_t >= batch_processor.buffer_duration then
+        core.log.debug(fmt("BatchProcessor[%s] buffer duration exceeded, activating buffer flush", batch_processor.name))
```

Comment:
code style: this line is too long

The single line of code should be less than 80 characters, and please fix other similar points.

Reply:
Refactored similar lines.